File: chad/src/main/java/org/woahoverflow/chad/commands/admin/ModifyPresence.java
package org.woahoverflow.chad.commands.admin;

import org.woahoverflow.chad.core.ChadBot;
import org.woahoverflow.chad.core.ChadVar;
import org.woahoverflow.chad.handle.MessageHandler;
import org.woahoverflow.chad.handle.commands.Command;
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageReceivedEvent;
import sx.blah.discord.handle.obj.ActivityType;
import sx.blah.discord.handle.obj.StatusType;

import java.util.HashMap;
import java.util.List;

public class ModifyPresence implements Command.Class {

    @Override
    public Runnable run(MessageReceivedEvent e, List<String> args) {
        return() -> {
            if (args.size() == 0)
            {
                new MessageHandler(e.getChannel()).sendError("Invalid Arguments");
                return;
            }
            String option = args.get(0);
            String message;
            switch (option) {
                default:
                    StringBuilder sb = new StringBuilder();
                    for (String str : args) {
                        sb.append(str).append(" ");
                    }
                    ChadBot.cli.changePresence(ChadVar.STATUS_TYPE, ActivityType.PLAYING, sb.toString().trim());
                    message = "Changed presence to \"" + sb.toString().trim() + "\"";
                    ChadVar.CURRENT_STATUS = sb.toString().trim();
                    break;
                case "rotate":
                    ChadVar.ROTATE_PRESENCE = true;
                    message = "Enabled presence rotation.";
                    break;
                case "static":
                    ChadVar.ROTATE_PRESENCE = false;
                    message = "Disabled presence rotation.";
                    break;
                case "add":
                    args.remove(0);
                    StringBuilder add_sb = new StringBuilder();
                    for (String str : args) {
                        add_sb.append(str).append(" ");
                    }
                    ChadVar.PRESENCE_ROTATION.add(add_sb.toString().trim());
                    message = "Added \"" + add_sb.toString().trim() + "\" to rotation";
                    break;
                case "time":
                    args.remove(0);
                    StringBuilder time_sb = new StringBuilder();
                    for (String str : args) {
                        time_sb.append(str).append(" ");
                    }
                    ChadVar.ROTATION_TIME = Integer.parseInt(time_sb.toString().trim());
                    message = "Changed rotation period to `" + ChadVar.ROTATION_TIME + "`";
                    break;
                case "status":
                    args.remove(0);
                    if (args.get(0).equalsIgnoreCase("idle"))
                    {
                        ChadVar.STATUS_TYPE = StatusType.IDLE;
                        e.getClient().changePresence(StatusType.IDLE, ActivityType.PLAYING, ChadVar.CURRENT_STATUS);
                        message = "Changed status type to `Idle`";
                        break;
                    }
                    else if (args.get(0).equalsIgnoreCase("online"))
                    {
                        ChadVar.STATUS_TYPE = StatusType.ONLINE;
                        e.getClient().changePresence(StatusType.ONLINE, ActivityType.PLAYING, ChadVar.CURRENT_STATUS);
                        message = "Changed status type to `Online`";
                        break;
                    }
                    else if (args.get(0).equalsIgnoreCase("offline"))
                    {
                        ChadVar.STATUS_TYPE = StatusType.OFFLINE;
                        e.getClient().changePresence(StatusType.INVISIBLE, ActivityType.PLAYING, ChadVar.CURRENT_STATUS);
                        message = "Changed status type to `Offline`";
                        break;
                    }
                    else if (args.get(0).equalsIgnoreCase("dnd"))
                    {
                        ChadVar.STATUS_TYPE = StatusType.DND;
                        e.getClient().changePresence(StatusType.DND, ActivityType.PLAYING, ChadVar.CURRENT_STATUS);
                        message = "Changed status type to `Do Not Disturb`";
                        break;
                    }
                    else {
                        message = "Invalid";
                    }
            }

            new MessageHandler(e.getChannel()).sendMessage(message);
        };
    }

    @Override
    public Runnable help(MessageReceivedEvent e) {
        HashMap<String, String> st = new HashMap<>();
        st.put("modpresence <string>", "Changes the bots rich presence message.");
        st.put("modpresence status <status>", "Changes the bots status.");
        st.put("modpresence <static/rotate>", "Disables or enables the bot's presence rotation.");
        return Command.helpCommand(st, "Modify Presence", e);
    }
}

File: chad/src/main/java/dev/shog/chad/commands/developer/ModifyPresence.kt
package dev.shog.chad.commands.developer

import dev.shog.chad.core.ChadVar
import dev.shog.chad.core.getClient
import dev.shog.chad.framework.handle.GuildHandler
import dev.shog.chad.framework.handle.MessageHandler
import dev.shog.chad.framework.handle.PresenceHandler
import dev.shog.chad.framework.obj.Command
import dev.shog.chad.framework.obj.Guild
import sx.blah.discord.handle.impl.events.guild.channel.message.MessageEvent
import sx.blah.discord.handle.obj.ActivityType
import sx.blah.discord.handle.obj.StatusType
import java.util.*

/**
 * Modify Chad's discord presence
 *
 * @author sho, codebasepw
 */
class ModifyPresence : Command.Class {
    override suspend fun run(e: MessageEvent, args: MutableList<String>) {
        val prefix = GuildHandler.getGuild(e.guild.longID).getObject(Guild.DataType.PREFIX) as String

        // Checks if there's no arguments
        if (args.isEmpty()) {
            MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, prefix + "modpresence [rotate/static/add/view/status/activity/new presence]")
            return
        }

        when (args[0]) {
            "refresh" -> {
                PresenceHandler.refreshPresences()
                MessageHandler(e.channel, e.author).sendMessage("Refreshed presences.")
            }

            "rotate" -> {
                ChadVar.rotatePresence = true
                MessageHandler(e.channel, e.author).sendMessage("Enabled presence rotation.")
            }

            "static" -> {
                ChadVar.rotatePresence = false
                MessageHandler(e.channel, e.author).sendMessage("Disabled presence rotation.")
            }

            "view" -> {
                val stringBuilder = StringBuilder()

                for (s in PresenceHandler.presences) stringBuilder.append("Activity Type: `${s.activityType}`, Status Type: `${s.statusType}`, Status: `${s.status}`\n")

                MessageHandler(e.channel, e.author).sendMessage(stringBuilder.removeSuffix("\n").toString())
            }

            "random" -> {
                PresenceHandler.randomPresence()

                MessageHandler(e.channel, e.author).sendMessage("Changed to random presence!")
            }

            "status" -> {
                // Removes the option argument
                args.removeAt(0)

                ChadVar.statusType = when (args[0].toLowerCase()) {
                    "idle" -> StatusType.IDLE
                    "online" -> StatusType.ONLINE
                    "offline" -> StatusType.INVISIBLE
                    "dnd" -> StatusType.DND
                    else -> {
                        MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "modpresence status [dnd/offline/online/idle]")
                        return
                    }
                }

                e.client.changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                MessageHandler(e.channel, e.author).sendMessage("Changed status type to `${ChadVar.statusType.name}`")
                return
            }

            "activity" -> {
                // Removes the option argument
                args.removeAt(0)

                ChadVar.activityType = when (args[0].toLowerCase()) {
                    "playing" -> ActivityType.PLAYING
                    "listening" -> ActivityType.LISTENING
                    "watching" -> ActivityType.WATCHING
                    "streaming" -> ActivityType.STREAMING
                    else -> {
                        MessageHandler(e.channel, e.author).sendPresetError(MessageHandler.Messages.INVALID_ARGUMENTS, "modpresence activity [playing/listening/watching]")
                        return
                    }
                }

                e.client.changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                MessageHandler(e.channel, e.author).sendMessage("Changed activity type to `${ChadVar.activityType.name}`")
                return
            }

            else -> {
                val sb = StringBuilder()
                for (arg in args) sb.append("$arg ")

                ChadVar.currentStatus = sb.toString().trim { it <= ' ' }
                getClient().changePresence(ChadVar.statusType, ChadVar.activityType, ChadVar.currentStatus)
                MessageHandler(e.channel, e.author).sendMessage("Changed presence to `${ChadVar.currentStatus}`.")
            }
        }
    }

    override suspend fun help(e: MessageEvent) {
        val st = HashMap<String, String>()
        st["modpresence activity [playing/listening/watching]]"] = "Changes the bot's activity."
        st["modpresence [string]"] = "Changes the bots rich presence message."
        st["modpresence status [dnd/offline/online/idle]"] = "Changes the bots status."
        st["modpresence [static/rotate]"] = "Disables or enables the bot's presence rotation."
        Command.helpCommand(st, "Modify Presence", e)
    }
}
